Source/FieldSolver/SpectralSolver/SpectralAlgorithms/GalileanPsatdAlgorithmRZ.cpp

```diff
@@ -0,0 +1,356 @@
+        int const modes = f.n_rz_azimuthal_modes;
+        amrex::ParallelFor(bx, modes,
+        [=] AMREX_GPU_DEVICE(int i, int j, int k, int mode) noexcept
+        {
+
+            // All of the fields of each mode are grouped together
+            using Idx = SpectralFieldIndex;
+            auto const Ep_m = Idx::Ex + Idx::n_fields*mode;
+            auto const Em_m = Idx::Ey + Idx::n_fields*mode;
+            auto const Ez_m = Idx::Ez + Idx::n_fields*mode;
+            auto const Bp_m = Idx::Bx + Idx::n_fields*mode;
+            auto const Bm_m = Idx::By + Idx::n_fields*mode;
+            auto const Bz_m = Idx::Bz + Idx::n_fields*mode;
+            auto const Jp_m = Idx::Jx + Idx::n_fields*mode;
+            auto const Jm_m = Idx::Jy + Idx::n_fields*mode;
+            auto const Jz_m = Idx::Jz + Idx::n_fields*mode;
+            auto const rho_old_m = Idx::rho_old + Idx::n_fields*mode;
+            auto const rho_new_m = Idx::rho_new + Idx::n_fields*mode;
+
+            // Record old values of the fields to be updated
+            Complex const Ep_old = fields(i,j,k,Ep_m);
+            Complex const Em_old = fields(i,j,k,Em_m);
+            Complex const Ez_old = fields(i,j,k,Ez_m);
+            Complex const Bp_old = fields(i,j,k,Bp_m);
+            Complex const Bm_old = fields(i,j,k,Bm_m);
+            Complex const Bz_old = fields(i,j,k,Bz_m);
+            // Shortcut for the values of J and rho
+            Complex const Jp = fields(i,j,k,Jp_m);
+            Complex const Jm = fields(i,j,k,Jm_m);
+            Complex const Jz = fields(i,j,k,Jz_m);
+            Complex const rho_old = fields(i,j,k,rho_old_m);
+            Complex const rho_new = fields(i,j,k,rho_new_m);
+
+            // k vector values, and coefficients
+            // The k values for each mode are grouped together
+            int const ir = i + nr*mode;
+            amrex::Real const kr = kr_arr[ir];
+            amrex::Real const kz = modified_kz_arr[j];
+
+            constexpr amrex::Real c2 = PhysConst::c*PhysConst::c;
+            constexpr amrex::Real inv_ep0 = 1._rt/PhysConst::ep0;
+            Complex const I = Complex{0._rt,1._rt};
+            amrex::Real const C = C_arr(i,j,k,mode);
+            amrex::Real const S_ck = S_ck_arr(i,j,k,mode);
+            Complex const X1 = X1_arr(i,j,k,mode);
+            Complex const X2 = X2_arr(i,j,k,mode);
+            Complex const X3 = X3_arr(i,j,k,mode);
+            Complex const X4 = X4_arr(i,j,k,mode);
+            Complex const T2 = Theta2_arr(i,j,k,mode);
+
+            // Update E (see WarpX online documentation: theory section)
+            fields(i,j,k,Ep_m) = T2*C*Ep_old
+                        + T2*S_ck*(-c2*I*kr/2._rt*Bz_old + c2*kz*Bp_old)
+                        + X4*Jp + kr*(X2*rho_new - T2*X3*rho_old);
+            fields(i,j,k,Em_m) = T2*C*Em_old
+                        + T2*S_ck*(-c2*I*kr/2._rt*Bz_old - c2*kz*Bm_old)
+                        + X4*Jm - kr*(X2*rho_new - T2*X3*rho_old);
```

Comment:
By looking at equation (C4b) of ([Lehe et al, PRE, 2016](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.94.053305)), it seems to me that we should have a factor 1/2 here. Could you crosscheck?
```suggestion
                        + X4*Jm - 0.5_rt*kr*(X2*rho_new - T2*X3*rho_old);
```